### ISSUE TITLE
[no-empty-function] Allow empty functions for protected class functions

### ISSUE THREAD

**Issue by clshortfuse** [typescript-eslint/typescript-eslint#5858]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-empty-function

### Description

We allow override methods to be empty with an option, but not the inverse (base class). See Pass/Fail for an example.

### Fail

```typescript
class Base {
  constructor() {
    document.getElementById('id')?.addEventListener('click', this.onClick);
  };
  onClick(event: Event) {};
}

class Extender extends Base {
  override onClick(event: Event) { console.log(event) }
}
```

### Pass

```typescript
class Base {
  constructor() {
    document.getElementById('id')?.addEventListener('click', this.onClick);
  };
  protected onClick(event: Event) {};
}

class Extender extends Base {
  override onClick(event: Event) { console.log(event) }
}
```

### Additional Info

Essential, the inverse of this issue:

https://github.com/typescript-eslint/typescript-eslint/issues/4919

**Comment by bradzacher** [comment]
I don't agree that this is a usecase that should be supported.

An `override` method makes sense to be empty - you **have** to provide an implementation. So it's actually very common to have an empty overriden function for cases where you don't want specific functionality.

OTOH there's no specific reason that a protected method must be empty. It's a rare case and a conscious implementation detail - it's an exception to the rule.
These sorts of cases are not things that should be built into the lint rule, instead they should be explicitly ignored with a disable comment and an explanation marking them clearly as an intentional empty function.

Otherwise it's very easy to end up with empty and unused cruft in your codebase (again a problem not present with an `override` function).

**Comment by bradzacher** [comment]
I don't agree that this is a usecase that should be supported.

An `override` method makes sense to be empty - you **have** to provide an implementation. So it's actually very common to have an empty overriden function for cases where you don't want specific functionality.

OTOH there's no specific reason that a protected method must be empty. It's a rare case and a conscious implementation detail - it's an exception to the rule.
These sorts of cases are not things that should be built into the lint rule, instead they should be explicitly ignored with a disable comment and an explanation marking them clearly as an intentional empty function.

If this was allowed in the rule then it would be very, very easy to accidentally end up with empty and unused protected function in your codebase (again a problem not present with an `override` function thanks to TS checks).

For the above reasons I don't think this is a good addition to the rule.
